components/num-util/src/fixed_decimal.rs

```diff
@@ -373,6 +376,154 @@ impl fmt::Display for FixedDecimal {
+            return Err(Error::Limit);
+        }
+        dec.upper_magnitude = temp_upper_magnitude as i16;
+
+        // Computing DecimalFixed.lower_magnitude
+        // Note: ((i16::MIN as u16) as usize) == 32768
+        let temp_lower_magnitude = no_dot_str_len - dot_index;
+        if temp_lower_magnitude > (i16::MIN as u16) as usize {
+            return Err(Error::Limit);
+        }
+        dec.lower_magnitude = (temp_lower_magnitude as i16).wrapping_neg();
+
+        // leftmost_digit: index of the first non-zero digit
+        // rightmost_digit: index of the first element after the last non-zero digit
+        // Example:
+        //     input string    leftmost_digit     rightmost_digit
+        //     00123000              2                  5
+        //     0.0123000             3                  6
+        //     001.23000             2                  6
+        //     001230.00             2                  5
+        // Compute leftmost_digit
+        let mut leftmost_digit = no_sign_str_len;
+        for (i, c) in no_sign_str.bytes().enumerate() {
+            if c == b'.' {
+                continue;
+            }
+            if c != b'0' {
+                leftmost_digit = i;
+                break;
+            }
+        }
+
+        // If the input only has zeros (like 000, 00.0, -00.000) we handle the situation here
+        // by returning the dec and don't running the rest of the code
+        if leftmost_digit == no_sign_str_len {
+            return Ok(dec);
+        }
+
+        // Else if the input is not all zeros, we compute its magnitude:
+        // Note that we can cast with "as" here because lower and upper magnitude have been checked already
+        let mut temp_magnitude = ((dot_index as i32) - (leftmost_digit as i32) - 1i32) as i16;
+        if dot_index < leftmost_digit {
+            temp_magnitude += 1;
+        }
+        dec.magnitude = temp_magnitude;
+
+        // Compute rightmost_digit
+        let mut rightmost_digit = no_sign_str_len;
+        for (i, c) in no_sign_str.bytes().rev().enumerate() {
+            if c == b'.' {
+                continue;
+            }
+            if c != b'0' {
+                rightmost_digit = no_sign_str_len - i;
+                break;
+            }
+        }
+
+        // digits_str_len: shows the length of digits (Ex. 0012.8900 --> 4)
+        let mut digits_str_len = rightmost_digit - leftmost_digit;
+        if leftmost_digit < dot_index && dot_index < rightmost_digit {
+            digits_str_len -= 1;
+        }
+
+        // Constructing DecimalFixed.digits
+        let mut v: SmallVec<[u8; 8]> = SmallVec::with_capacity(digits_str_len);
+        for (i, c) in no_sign_str[leftmost_digit..rightmost_digit]
+            .bytes()
+            .enumerate()
+        {
+            if c == b'.' {
+                continue;
+            }
+            if i > std::i16::MAX as usize {
+                return Err(Error::Limit);
+            }
```

Comment:
It's okay if `i` exceeds `i16::MAX`, because the digit list can be up to `u16::MAX` digits long.  You don't need to worry about checking anything here because `upper_magnitude` and `lower_magnitude` are already in range.